docs/2019-05-23-Unification-Modulo-Equations.md

```diff
@@ -0,0 +1,163 @@
+in a way that can be used almost directly by the Haskell backend (some
+changes needed).
+
+These axioms would allow us to reduce the hard to handle
+unification cases to unambiguous ones (e.g. reduce AC unification to
+constructor-based unification).
+
+See the use-cases below for examples and explanations.
+
+We could write these in a way that we don't duplicate unification cases.
+Then we could apply these as simplification axioms, defining our unification
+strategy in k. When we need performance, we could implement the ones that are
+used more frequently in Haskell.
+
+Or, if that ever becomes desirable (e.g. if we want it to make proof
+certificates easier to generate), we could give up on performance and define
+all simplification cases as axioms/theorems.
+
+Also, we should probably use set variables, to avoid the large number of
+checks needed when working with normal variables (though, of course, we
+need to make sure that the axioms still hold).
+
+Use-case: lists
+---------------
+
+For lists, we could have the following axioms
+```
+⌈elem(x)∧unit⌉ = ⊥
+⌈elem(x)∧elem(y)⌉ = ⌈x∧y⌉
+⌈concat(x,y)∧unit⌉ = ⌈x∧unit⌉ ∧ ⌈y∧unit⌉
+⌈concat(elem(x),y) ∧ concat(elem(z),t)⌉ = ⌈x∧z⌉ ∧ ⌈y∧t⌉
+```
+This is missing some axioms (e.g. the ones unifying at the end of the list,
+or splitting the list), but it looks like these, together with the usual list
+axioms and the `and` commutativity axioms could be enough to solve list
+unification.
+
+Note that applying these axioms allows us to reduce complex expressions to
+simpler ones, without worrying about how many terms are equivalent to the
+one we have: we just have to bring the term in one of the forms required for
+one of these axioms, then we apply the axiom and that’s it.
+
+As an example, say we need to unify `concat(concat(elem(φ_1), elem(φ_2)), φ_3)`
+with `concat(concat(unit, elem(ψ_1)), concat(elem(ψ_2), ψ_3))`. We first use
+the list axioms to rewrite the first term as
+`concat(elem(φ_1), concat(elem(φ_2), φ_3))`, the second one as
+`concat(elem(ψ_1), concat(elem(ψ_2), ψ_3))`, then we apply the
+`⌈concat(elem(x),y) ∧ concat(elem(z),t)⌉ = ⌈x∧z⌉ ∧ ⌈y∧t⌉` axiom repeatedly
+to compute
+```
+⌈ concat(elem(φ_1), concat(elem(φ_2), φ_3))
+∧ concat(elem(ψ_1), concat(elem(ψ_2), ψ_3)) ⌉
+    = ⌈φ_1∧ψ_1⌉ ∧ ⌈concat(elem(φ_2), φ_3) ∧ concat(elem(ψ_2), ψ_3)⌉
+    = ⌈φ_1∧ψ_1⌉ ∧ ⌈φ_2∧ψ_2⌉ ∧ ⌈φ_3∧ψ_3⌉
+```
+
+Use-case: sets
+--------------
+
+Let us try to define the axioms for set lookup through unification, i.e.
+for unifying `concat(elem(x), y)` with a set. Assuming that our terms are
+normalized (elems moved to the left, concats represented as
+`concat(a, concat(b, concat(...)))`), then the following axioms should be
+enough:
+
+```
+⌈concat(x, y) ∧ unit⌉ = ⌈x∧unit⌉ ∧ ⌈y∧unit⌉
+⌈concat(elem(x), y) ∧ elem(a)⌉ = ⌈x∧a⌉ ∧ ⌈y∧unit⌉
+⌈concat(elem(x), y) ∧ concat(elem(a), b)⌉
+   = ⌈x∧a⌉ ∧ ⌈y∧b⌉
+     ∨ ∃ z . ⌈concat(elem(x),z)∧b⌉ ∧ ⌈y∧concat(elem(a),z)⌉
+```
```

Comment:
How about sets (not multi-sets) where `concat(x,x) = x`? What would the last axiom look like?

Reply:
So, I thought that, for maps, `concat(elem(a, b), elem(a, ?))` is bottom. I assumed that, for sets, it's the same. Now I realize that while `concat(elem(a, b), elem(a, c))` is probably bottom, `concat(elem(a, b), elem(a, b))` might actually be `elem(a, b)`. So, just to make sure, how is this supposed to work?

Reply:
I don't know actually :) That's why I ask. The axiom in line 127 looks very interesting, and it takes me quite a while to digest its intuition.

Reply:
K's `Set` has the property that `concat(X, X) = \bottom` so that `concat` is commutative if its arguments are distinct, but share keys.

Reply:
Added a section for sets.